golang/cosmos/app/app.go

```diff
@@ -809,57 +824,87 @@ func normalizeModuleAccount(ctx sdk.Context, ak authkeeper.AccountKeeper, name s
 type cosmosInitAction struct {
 	Type        string             `json:"type"`
 	ChainID     string             `json:"chainID"`
+	IsBootstrap bool               `json:"isBootstrap"`
 	Params      swingset.Params    `json:"params"`
-	StoragePort int                `json:"storagePort"`
 	SupplyCoins sdk.Coins          `json:"supplyCoins"`
-	VibcPort    int                `json:"vibcPort"`
-	VbankPort   int                `json:"vbankPort"`
-	LienPort    int                `json:"lienPort"`
 	UpgradePlan *upgradetypes.Plan `json:"upgradePlan,omitempty"`
+	LienPort    int                `json:"lienPort"`
+	StoragePort int                `json:"storagePort"`
+	VbankPort   int                `json:"vbankPort"`
+	VibcPort    int                `json:"vibcPort"`
 }
 
 // Name returns the name of the App
 func (app *GaiaApp) Name() string { return app.BaseApp.Name() }
 
-func (app *GaiaApp) MustInitController(ctx sdk.Context) {
-	if app.controllerInited {
-		return
+func (app *GaiaApp) CheckControllerInited(expected bool) {
+	if app.controllerInited != expected {
+		fmt.Fprintf(os.Stderr, "controllerInited != %t\n", expected)
+		debug.PrintStack()
+		os.Exit(1)
 	}
+}
+
+func (app *GaiaApp) initController(ctx sdk.Context, bootstrap bool) {
+	app.CheckControllerInited(false)
 	app.controllerInited = true
 	// Begin initializing the controller here.
 	action := &cosmosInitAction{
 		Type:        "AG_COSMOS_INIT",
 		ChainID:     ctx.ChainID(),
+		IsBootstrap: bootstrap,
 		Params:      app.SwingSetKeeper.GetParams(ctx),
-		StoragePort: app.vstoragePort,
 		SupplyCoins: sdk.NewCoins(app.BankKeeper.GetSupply(ctx, "uist")),
-		VibcPort:    app.vibcPort,
-		VbankPort:   app.vbankPort,
-		LienPort:    app.lienPort,
 		UpgradePlan: app.upgradePlan,
+		LienPort:    app.lienPort,
+		StoragePort: app.vstoragePort,
+		VbankPort:   app.vbankPort,
+		VibcPort:    app.vibcPort,
 	}
+	// This really abuses `BlockingSend` to get back at `sendToController`
 	out, err := app.SwingSetKeeper.BlockingSend(ctx, action)
 
 	// fmt.Fprintf(os.Stderr, "AG_COSMOS_INIT Returned from SwingSet: %s, %v\n", out, err)
 
 	if err != nil {
-		fmt.Fprintln(os.Stderr, "Cannot initialize Controller", err)
-		os.Exit(1)
+		panic(fmt.Errorf("cannot initialize Controller, %w", err))
```

Comment:
@michaelfig would there be a way to allow these `%w` in errors used for panic? I'd like to keep the stack trace in this case.

Reply:
Untested, but. this should be equivalent:

```go
import (
	"github.com/cosmos/cosmos-sdk/types/errors"
)
```

```suggestion
		panic(errors.Wrap(err, "cannot initialize Controller"))
```